crates/core/src/subscription/module_subscription_actor.rs

```diff
@@ -47,44 +42,49 @@ impl ModuleSubscriptions {
         });
 
         let auth = AuthCtx::new(self.owner_identity, sender.id.identity);
-        let mut n_queries = 0;
-        let queries = subscription
-            .query_strings
-            .iter()
-            .map(|sql| compile_read_only_query(&self.relational_db, &tx, &auth, sql))
-            .flatten_ok()
-            .inspect(|_| n_queries += 1)
-            .collect::<Result<ExecutionSet, _>>()?;
-
-        let database_update = queries.eval(&self.relational_db, &tx, auth)?;
+        let mut queries = vec![];
+
+        let guard = self.subscriptions.read();
+
+        for sql in subscription.query_strings {
+            let hash = QueryHash::from_string(&sql);
+            if let Some(unit) = guard.query(&hash) {
+                queries.push(unit);
+            } else {
+                queries.extend(
+                    compile_read_only_query(&self.relational_db, &tx, &auth, &sql)?
+                        .into_iter()
+                        .map(|plan| Arc::new(ExecutionUnit::new(plan, hash))),
+                )
```

Comment:
Why do we have `queries.push` in one branch, but `queries.extend` in the other? Surely we want to add exactly 1 element in either branch, right?

Reply:
Correct, this assumes that compile_read_only_query returns a single element vec. I should either add an assert, or change the type signature.

Reply:
I would also accept a comment to that effect. Any of the three would be fine, just so long as a reader can tell what's going on.